test/cpp/naming/naming_end2end_test.cc

```diff
@@ -0,0 +1,297 @@
+    gpr_mu_unlock(args->mu);
+    grpc_exec_ctx_finish(&exec_ctx);
+  }
+  gpr_event_set(&args->ev, (void *)1);
+}
+
+static void check_service_config_result_locked(grpc_channel_args *channel_args,
+                                               args_struct *args) {
+  const grpc_arg *service_config_arg =
+      grpc_channel_args_find(channel_args, GRPC_ARG_SERVICE_CONFIG);
+  if (args->expected_service_config_string != NULL &&
+      strlen(args->expected_service_config_string) > 0) {
+    GPR_ASSERT(service_config_arg != NULL);
+    GPR_ASSERT(service_config_arg->type == GRPC_ARG_STRING);
+    char *service_config_string = service_config_arg->value.string;
+    if (gpr_stricmp(service_config_string,
+                    args->expected_service_config_string) != 0) {
+      gpr_log(GPR_ERROR, "expected service config string: |%s|",
+              args->expected_service_config_string);
+      gpr_log(GPR_ERROR, "got service config string: |%s|",
+              service_config_string);
+      GPR_ASSERT(0);
+    }
+  } else {
+    GPR_ASSERT(service_config_arg == NULL);
+  }
+}
+
+static void check_channel_arg_srv_result_locked(grpc_exec_ctx *exec_ctx,
+                                                void *argsp, grpc_error *err) {
+  args_struct *args = (args_struct *)argsp;
+  grpc_channel_args *channel_args = args->channel_args;
+  const grpc_arg *channel_arg =
+      grpc_channel_args_find(channel_args, GRPC_ARG_LB_ADDRESSES);
+  GPR_ASSERT(channel_arg != NULL);
+  GPR_ASSERT(channel_arg->type == GRPC_ARG_POINTER);
+  grpc_lb_addresses *addresses =
+      (grpc_lb_addresses *)channel_arg->value.pointer.p;
+  gpr_log(GPR_INFO, "num addrs found: %d. expected %" PRIdPTR,
+          (int)addresses->num_addresses, args->expected_addrs.size());
+
+  GPR_ASSERT(addresses->num_addresses == args->expected_addrs.size());
+  for (size_t i = 0; i < addresses->num_addresses; i++) {
+    grpc_lb_address addr = addresses->addresses[i];
+    char *str;
+    grpc_sockaddr_to_string(&str, &addr.address, 1 /* normalize */);
+    gpr_log(GPR_INFO, "%s", str);
+    GPR_ASSERT(addr.is_balancer == args->expect_is_balancer);
+    GPR_ASSERT(matches_any(args->expected_addrs, str));
+    gpr_free(str);
+  }
+
+  check_service_config_result_locked(channel_args, args);
+
+  gpr_atm_rel_store(&args->done_atm, 1);
+  gpr_mu_lock(args->mu);
+  GRPC_LOG_IF_ERROR("pollset_kick", grpc_pollset_kick(args->pollset, NULL));
+  gpr_mu_unlock(args->mu);
+}
+
+static void test_resolves(grpc_exec_ctx *exec_ctx, args_struct *args) {
+  char *whole_uri = NULL;
+  const char *authority = FLAGS_local_dns_server_address.c_str();
+  if (authority != NULL && strlen(authority) > 0) {
+    gpr_log(GPR_INFO, "Specifying authority in uris to: %s", authority);
+  } else {
+    authority = "";
+  }
+
+  GPR_ASSERT(asprintf(&whole_uri, "dns://%s/%s", authority,
+                      (char *)args->target_name));
+  args->channel_args = NULL;
```

Comment:
Please move this into `naming_end2end_test_resolution()`, where the rest of the args struct is initialized.